### PR TITLE
Do not emit unnecessary fields for nested objects

### LINKED ISSUE #3950: Unused field generated for nested objects
```scala
object Foo {
  object Bar
}
```
Decompiled:
```java
public final class Foo$ {
    public static final Foo$ MODULE$;
    public final Foo.Bar$ Bar; // Not here with scalac

    public static {
        new Foo$();
    }

    public Foo$() {
        MODULE$ = this;
    }
}
```

### PR DESCRIPTION
**Pull request by SolalPirelli**
Fixes #3950

## How much have your relied on LLM-based tools in this contribution?

not

## How was the solution tested?

new automated test

### CODE REVIEW

**Comment by SethTisue** [comment]
Perhaps this should be labeled `needs-next-minor`? Although we hope the unnecessary fields aren't being used by anyone, it could still be prudent not to do it in a patch version?

I imagine MiMa will notice the missing fields?

**Comment by SolalPirelli** [comment]
Yeah I think the question is (1) what minor and (2) should they be first emitted as `@Deprecated`

**Comment by SolalPirelli** [comment]
OK so as just discussed at Core the fact they're never initialized makes this easier, we'll do it for the next minor.

File: compiler/src/dotty/tools/backend/jvm/BCodeSkelBuilder.scala [hunk 1]
```diff
@@ -405,15 +405,8 @@ trait BCodeSkelBuilder extends BCodeHelpers {
...
     }
 
-    def addClassFields()(using Context): Unit =
-      /*  Non-method term members are fields, except for module members. Module
-       *  members can only happen on .NET (no flatten) for inner traits. There,
-       *  a module symbol is generated (transformInfo in mixin) which is used
-       *  as owner for the members of the implementation class (so that the
-       *  backend emits them as static).
-       *  No code is needed for this module symbol.
-       */
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/backend/jvm/BCodeSkelBuilder.scala:416]: This comment seemed very obsolete

File: tests/pos-macros/i12788/Test.scala [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// nopos-error
+// it's also OK if this errors due to a cycle, what matters is the compiler shouldn't crash
```
Comment (SolalPirelli) [inline tests/pos-macros/i12788/Test.scala:1]: per Jan on Slack:
> Macro.scala needs to be compiled before Test.scala, because of the macro call caused by lhs.width. So Test.scala gets suspended, and we only have access to the stale DFBitsCompanion symbol which will not cause issues if it's not accessed, so I suspect that's what happened in your PR (it's no longer accessed). Whether it should fail or succeed in these corner cases is kind of arbitrary with how the suspension mechanism is implemented (it's reactive, not proactive), so I think it would be fine to move this to a pos test (maybe with some comment) - the important part for this is not to cause any crashes/other nonsense errors

Comment (lrytz) [reply tests/pos-macros/i12788/Test.scala:1]: I don't really understand what "OK if this errors due to a cycle" means here..

**Review by lrytz — APPROVED** [review]

File: tests/pos-macros/i12788/Test.scala (hunk 2 above)
Comment (SolalPirelli) [reply tests/pos-macros/i12788/Test.scala:1]: before this PR this was in neg due to a cyclic reference error
